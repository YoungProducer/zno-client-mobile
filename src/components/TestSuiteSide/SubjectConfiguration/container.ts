/**
 * Created by: Oleksandr Bezrukov
 * Creation date: 4 March 2020
 *
 * Declare main types and interfaces for SubjectConfiguration component.
 * Create function which connect actions
 * and/or variables from the redux-store to the component.
 */

/** External imports */
import { connect } from 'react-redux';

/** Application's imports */
import { fetchSubjectConfigurationAction } from 'store/actionCreators/subjectConfiguration';
import {
    toggleSubjectConfigurationDialogAction, ISubjectRootData,
} from 'store/slices/subjectConfiguration';
import {
    selectSubjectConfigurationLoading,
    selectSubjectConfigurationDialogVisible,
    selectSubjectConfigSubjectData,
    selectSubjectConfigThemes,
    selectSubjectConfigExams,
    selectSubjectConfigSubSubjectsThemes,
    selectSubjectConfigSubSubjectsData,
} from 'store/selectors/subjectConfiguration';
import { selectIsLoggedIn } from 'store/selectors/auth/signin';
import { RootState } from 'store/slices';
import { ISubjectConfigurationCredentials } from 'api';

/** Define types for test configuration */
export const enum ETestTypes {
    THEMES = 'Вибір теми',
    EXAMS = 'Тренувальні варіанти ЗНО',
}

export const enum EExamTypes {
    SESSIONS = 'ОСНОВНА СЕССІЯ ЗНО',
    TRAININGS = 'ТРЕНУВАЛЬНІ ВАРІАНТИ',
}

/** Props which component get from the parent */
interface IOwnProps {}

/** Props which component select from the redux-store */
interface IStateProps {
    isLoggedIn: boolean;
    loading: boolean;
    dialogVisible: boolean;
    subjectData: ISubjectRootData;
    subjectThemes: string[] | null;
    subjectExams: {
        trainings: string[] | null;
        sessions: string[] | null;
    };
    subSubjectsData: ISubjectRootData[] | null;
    subSubjectsThemes: {
        [attr: string]: string[];
    };
}

/** Props(actions) which component can dispatch */
interface IDispatchProps {
    fetchSubjectConfiguration: (credentials: ISubjectConfigurationCredentials) => void;
    toggleSubjectConfigurationDialog: (value: boolean) => void;
}

/** Declare type which describe all props pushed to the component */
export type TSubjectConfigurationModalProps =
    IOwnProps
    & IStateProps
    & IDispatchProps;

/** Select variables from the redux-store */
const mapStateToProps = (state: RootState): IStateProps => ({
    isLoggedIn: selectIsLoggedIn(state),
    loading: selectSubjectConfigurationLoading(state),
    dialogVisible: selectSubjectConfigurationDialogVisible(state),
    subjectData: selectSubjectConfigSubjectData(state),
    subjectThemes: selectSubjectConfigThemes(state),
    subjectExams: selectSubjectConfigExams(state),
    subSubjectsData: selectSubjectConfigSubSubjectsData(state),
    subSubjectsThemes: selectSubjectConfigSubSubjectsThemes(state),
});

/** Create functions which dispatch some actions */
const mapDispatchToProps = (dispatch: any): IDispatchProps => ({
    fetchSubjectConfiguration: (credentials: ISubjectConfigurationCredentials) =>
        dispatch(fetchSubjectConfigurationAction(credentials)),

    toggleSubjectConfigurationDialog: (value: boolean) =>
        dispatch(toggleSubjectConfigurationDialogAction(value)),
});

/**
 * Export function which connect actions
 * and/or variables from the redux-store to the component.
 */
export default connect<IStateProps, IDispatchProps, IOwnProps>(
    mapStateToProps,
    mapDispatchToProps,
);
