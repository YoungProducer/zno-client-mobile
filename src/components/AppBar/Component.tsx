/**
 * Created by: Oleksandr Bezrukov
 * Creation date: 2 March 2020
 *
 * Header component which contain main routes and another information or actions.
 */

/** External imports */
import React from 'react';
import CssBaseLine from '@material-ui/core/CssBaseline';
import Container from '@material-ui/core/Container';
import AppBar from '@material-ui/core/AppBar';

/** Application's imports */
import { TAppBarProps } from './container';
import NavigationLink from 'components/NavigationLink';

/** Create component */
const Component = (props: TAppBarProps) => {
    const { classes } = props;

    return (
        <React.Fragment>
            <CssBaseLine />
            <AppBar
                position='static'
                className={classes.root}
            >
                <Container
                    maxWidth='lg'
                    className={classes.container}
                >
                    <NavigationLink
                        navLink={{
                            to: '/',
                        }}
                        className={classes.navLink}
                        data-testid='home-navlink'
                    >
                        Головна
                    </NavigationLink>
                    <NavigationLink
                        navLink={{
                            to: '/subject-selection',
                        }}
                        className={classes.navLink}
                        data-testid='tests-navlink'
                    >
                        Тести
                    </NavigationLink>
                </Container>
            </AppBar>
        </React.Fragment>
    );
};

export default Component;
