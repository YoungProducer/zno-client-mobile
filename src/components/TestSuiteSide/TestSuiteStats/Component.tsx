/**
 * Created by: Oleksandr Bezrukov
 * Creation date: 26 March 2020
 *
 * Component which display main stats for current test suite
 * and has some actions.
 */

/** External imports */
import React, { useContext, useEffect } from 'react';
import { useHistory } from 'react-router-dom';
import IconButton from '@material-ui/core/IconButton';
import HomeIcon from '@material-ui/icons/HomeOutlined';
import Typography from '@material-ui/core/Typography';
import { makeStyles, createStyles, Theme } from '@material-ui/core/styles';

/** Application's imports */
import { AdditionalAnswerPropertiesContext } from 'context/TestSuiteContext';
import Wrapper from '../Wrapper';
import { TTestSuiteStatsProps } from './container';
import Progress from 'components/custom/LinearProgress';
import Actions from './Actions';

const useStyles = makeStyles((theme: Theme) =>
    createStyles({
        title: {
            color: '#867272',
            lineHeight: '48px',
        },
        dialogHeader: {
            paddingLeft: theme.spacing(2),
            paddingRight: theme.spacing(2),
            height: 48,
            position: 'relative',
        },
        progressBlock: {
            marginTop: theme.spacing(2),
        },
        actions: {
            marginTop: theme.spacing(2),
        },
        icon: {
            color: '#333',
        },
        homeIcon: {
            position: 'absolute',
            left: 0,
        },
    }));

const Component = (props: TTestSuiteStatsProps) => {
    const classes = useStyles({});

    const {
        amountOfGived,
        amountOfRight,
        amountOfSelected,
        answersAmount,
        currentAmountOfPoints,
        maxAmountOfPoints,
        finished,
        setTestSuiteFinished,
    } = props;

    const history = useHistory();

    const { showRightDuringTest } = useContext(AdditionalAnswerPropertiesContext);

    const redirectToHome = () => history.push('/');

    const finishTestSuiteHandler = () => setTestSuiteFinished(true);

    return (
        <Wrapper>
            <div className={classes.dialogHeader}>
                <IconButton
                    onClick={redirectToHome}
                    className={classes.homeIcon}
                >
                    <HomeIcon className={classes.icon}/>
                </IconButton>
                <Typography
                    variant='h5'
                    align='center'
                    className={classes.title}
                >
                    Статистика тесту
                </Typography>
            </div>
            <div className={classes.progressBlock}>
                <Progress
                    current={amountOfSelected}
                    total={answersAmount}
                    label='Кількість завдань у яких вибрана відповідь'
                />
                <Progress
                    current={amountOfGived}
                    total={answersAmount}
                    label='Кількість завдань на які дана відповідь'
                />
                <Progress
                    current={amountOfRight}
                    total={answersAmount}
                    label='Кількість завдань на які дана правильна відповідь'
                    hide={!showRightDuringTest ? !finished : false}
                />
                <Progress
                    current={currentAmountOfPoints}
                    total={maxAmountOfPoints}
                    label='Кількість набраних балів'
                    hide={!showRightDuringTest ? !finished : false}
                />
            </div>
            <div className={classes.actions}>
                <Actions finishTest={finishTestSuiteHandler}/>
            </div>
        </Wrapper>
    );
};

export default Component;
