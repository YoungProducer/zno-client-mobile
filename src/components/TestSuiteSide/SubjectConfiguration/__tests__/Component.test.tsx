/**
 * Created by: Oleksandr Bezrukov
 * Creation date: 4 March 2020
 *
 * Create test suites for SubjectConfigurationModal component.
 * @jest-environment jsdom
 */

/** External imports */
import React from 'react';
import { shallow } from 'enzyme';

/** Application's imports */
import Component from '../Component';
import { ETestTypes, TSubjectConfigurationModalProps, EExamTypes } from '../container';

const push = jest.fn();

jest.mock('react-router-dom', () => ({
    useLocation: () => ({
        pathname: '/subject-selection',
    }),
    useHistory: jest.fn(() => ({
        push,
    })),
    useParams: () => ({
        subjectId: 'foo',
    }),
}));

describe('SubjectConfigurationModal component', () => {
    /** Create mocked functions for props */
    const fetchSubjectConfiguration = jest.fn();
    const toggleSubjectConfigurationDialog = jest.fn();

    /** Define required props */
    const requiredProps: TSubjectConfigurationModalProps = {
        fetchSubjectConfiguration,
        toggleSubjectConfigurationDialog,
        isLoggedIn: false,
        dialogVisible: true,
        loading: false,
        subjectData: {
            name: 'foo',
            id: '123',
        },
        subSubjectsData: null,
        subSubjectsThemes: null,
        subjectExams: null,
        subjectThemes: null,
    };

    beforeEach(() => {
        push.mockReset();
        (requiredProps.fetchSubjectConfiguration as jest.Mock).mockReset();
        (requiredProps.toggleSubjectConfigurationDialog as jest.Mock).mockReset();
    });

    test('Is match snapshot', () => {
        /** Render component in shallow via enzyme */
        const tree = shallow(<Component {...requiredProps} />);

        /** Assert component matches snapshot */
        expect(tree).toMatchSnapshot();
    });

    test('Select test type', () => {
        /** Render component */
        const tree = shallow(<Component {...requiredProps} />);

        /** Simulate change event */
        tree.find(`[data-testid='select-test-type']`).simulate('change', {
            target: {
                value: ETestTypes.THEMES,
            },
        });

        /** Assert testType has correct value */
        expect(tree.find(`[data-testid='select-test-type']`).props().value).toEqual(ETestTypes.THEMES);
    });

    test('Dialog close button handler', () => {
        /** Render component */
        const tree = shallow(<Component {...requiredProps} />);

        /** Simulate click on 'close' button */
        tree.find(`[data-testid='close-subject-conf-button']`)
            .simulate('click');

        /** Assert 'toggleSubjectConfigurationDialog' function has been called */
        expect(toggleSubjectConfigurationDialog).toHaveBeenCalled();
    });

    test(`If subSubjectsNames is not null and testType equals 'THEMES' component should render sub-subject selection text-field`, () => {
        /** Render component */
        const tree = shallow(
            <Component
                {...requiredProps}
                subSubjectsData={[{
                    name: 'foo',
                    id: '123',
                }]}
            />,
        );

        /** Assert sub-subject selection text-field doesn't exist */
        expect(tree.exists(`[data-testid='select-sub-subject']`)).toBeFalsy();

        /** Simulate onChange event in select-test-type text field */
        tree.find(`[data-testid='select-test-type']`).simulate('change', { target: { value: ETestTypes.THEMES } });

        /** Assert sub-subject selection text-field exists */
        expect(tree.exists(`[data-testid='select-sub-subject']`)).toBeTruthy();
    });

    test(`If subSubjectsNames is null select-sub-subject text field shouldn't exists`, () => {
        /** Render component */
        const tree = shallow(
            <Component
                {...requiredProps}
                subSubjectsData={null}
            />,
        );

        /** Assert sub-subject selection text-field doesn't exist */
        expect(tree.exists(`[data-testid='select-sub-subject']`)).toBeFalsy();

        /** Simulate onChange event in select-test-type text field */
        tree.find(`[data-testid='select-test-type']`).simulate('change', { target: { value: ETestTypes.THEMES } });

        /** Assert sub-subject selection text-field exists */
        expect(tree.exists(`[data-testid='select-sub-subject']`)).toBeFalsy();
    });

    test(`If subSubjectsNames is not null and testType equals 'THEMES' select sub-subject`, () => {
        /** Render component */
        const tree = shallow(
            <Component
                {...requiredProps}
                subSubjectsData={[{
                    name: 'foo',
                    id: '123',
                }]}
            />,
        );

        /** Assert sub-subject selection text-field doesn't exist */
        expect(tree.exists(`[data-testid='select-sub-subject']`)).toBeFalsy();

        /** Simulate onChange event in select-test-type text field */
        tree.find(`[data-testid='select-test-type']`).simulate('change', { target: { value: ETestTypes.THEMES } });

        /** Assert sub-subject selection text-field exists */
        expect(tree.exists(`[data-testid='select-sub-subject']`)).toBeTruthy();

        /** Simulate onChange event in select-sub-subject text-field */
        tree.find(`[data-testid='select-sub-subject']`).simulate('change', { target: { value: 'foo' } });

        /** Assert select-sub-subject has value 'foo' */
        expect(tree.find(`[data-testid='select-sub-subject']`).props().value).toEqual('foo');
    });

    test(`If subjectThemes is null select-subject-theme text field shouldn't exist`, () => {
        /** Render component */
        const tree = shallow(
            <Component
                {...requiredProps}
                subjectThemes={null}
                subSubjectsData={null}
            />,
        );

        /** Simulate onChange event in select-test-type text field */
        tree.find(`[data-testid='select-test-type']`).simulate('change', { target: { value: ETestTypes.THEMES } });

        /** Assert select-sub-themes selection doesn't exist */
        expect(tree.exists(`[data-testid='select-subject-theme']`)).toBeFalsy();
    });

    test(`If subjectThemes is not null and testType equals 'Themes' select-subject-theme text field should exists`, () => {
        /** Render component */
        const tree = shallow(
            <Component
                {...requiredProps}
                subjectThemes={['bar']}
            />,
        );

        /** Simulate onChange event in select-test-type text field */
        tree.find(`[data-testid='select-test-type']`).simulate('change', { target: { value: ETestTypes.THEMES } });

        /** Assert select-sub-themes selection exists */
        expect(tree.exists(`[data-testid='select-subject-theme']`)).toBeTruthy();
    });

    test(`If subjectThemes is not null and testType equals 'Themes' select theme`, () => {
        /** Render component */
        const tree = shallow(
            <Component
                {...requiredProps}
                subjectThemes={['bar']}
            />,
        );

        /** Simulate onChange event in select-test-type text field */
        tree.find(`[data-testid='select-test-type']`).simulate('change', { target: { value: ETestTypes.THEMES } });

        /** Assert select-subject-theme selection text-field exists */
        expect(tree.exists(`[data-testid='select-subject-theme']`)).toBeTruthy();

        /** Simulate onChange event in select-subject-theme text field */
        tree.find(`[data-testid='select-subject-theme']`).simulate('change', { target: { value: 'bar' } });

        /** Assert select-subject-theme text field has correct value */
        expect(tree.find(`[data-testid='select-subject-theme']`).props().value).toBe('bar');
    });

    test(`If subSubjectThemes is not null and testType equals 'Themes' component should display theme selection`, () => {
        /** Render component */
        const tree = shallow(
            <Component
                {...requiredProps}
                subSubjectsData={[{
                    name: 'foo',
                    id: '123',
                }]}
                subSubjectsThemes={{ foo: ['bar'] }}
            />,
        );

        /** Simulate onChange event in select-test-type text field */
        tree.find(`[data-testid='select-test-type']`)
            .simulate('change', { target: { value: ETestTypes.THEMES } });

        /** Assert select-subject-theme selection text-field exists */
        expect(tree.exists(`[data-testid='select-subject-theme']`)).toBeTruthy();
    });

    test(`If 'subjectThemes' exists after selection of test type, theme selection text field should set '0' el of array as default`, () => {
        /** Render component */
        const tree = shallow(
            <Component
                {...requiredProps}
                subjectThemes={['foo', 'bar']}
            />,
        );

        /** Simulate onChange event in select-test-type text field */
        tree.find(`[data-testid='select-test-type']`)
            .simulate('change', { target: { value: ETestTypes.THEMES } });

        /** Assert select-subject-theme text field has right value */
        expect(tree.find(`[data-testid='select-subject-theme']`).props().value).toBe('foo');
    });

    test(`If 'subSubjectThemes' exists, after selecto of the test type, theme selection text field should set '0' el of array as default value`, () => {
        /** Render component */
        const tree = shallow(
            <Component
                {...requiredProps}
                subSubjectsData={[{
                    name: 'foo',
                    id: '123',
                }]}
                subSubjectsThemes={{ foo: ['bar', 'abc'] }}
            />,
        );

        /** Simulate onChange event in select-test-type text field */
        tree.find(`[data-testid='select-test-type']`)
            .simulate('change', { target: { value: ETestTypes.THEMES } });

        /** Assert select-subject-theme text field has right value */
        expect(tree.find(`[data-testid='select-subject-theme']`).props().value).toBe('bar');
    });

    test(`If subSubjectName doesn't exist in subSubjectThemes`, () => {
        /** Render component */
        const tree = shallow(
            <Component
                {...requiredProps}
                subSubjectsData={[{
                    name: 'foo',
                    id: '123',
                }]}
                subSubjectsThemes={{ zoo: ['bar'] }}
            />,
        );

        /** Simulate onChange event in select-test-type text field */
        tree.find(`[data-testid='select-test-type']`)
            .simulate('change', { target: { value: ETestTypes.THEMES } });

        /** Assert select-subject-theme selection text-field exists */
        expect(tree.exists(`[data-testid='select-subject-theme']`)).toBeTruthy();
    });

    test(`If 'subjectExams' exists select exam type if test type equals 'THEMES'`, () => {
        /** Render component */
        const tree = shallow(
            <Component
                {...requiredProps}
                subjectExams={{
                    sessions: ['foo'],
                    trainings: ['bar'],
                }}
            />,
        );

        /** Simulate onChange event in select-test-type text field */
        tree.find(`[data-testid='select-test-type']`)
            .simulate('change', { target: { value: ETestTypes.EXAMS } });

        /** Assert select-test-type field has value 'EXAMS' */
        expect(tree.find(`[data-testid='select-test-type']`).props().value).toBe(ETestTypes.EXAMS);

        /** Assert select-exam-type field exists */
        expect(tree.exists(`[data-testid='select-exam-type']`)).toBeTruthy();

        /** Simulate onChange event in select-exam-type text field */
        tree.find(`[data-testid='select-exam-type']`)
            .simulate('change', { target: { value: EExamTypes.TRAININGS } });

        /** Assert select-exam-type field has value 'TRAININGS' */
        expect(tree.find(`[data-testid='select-exam-type']`).props().value).toBe(EExamTypes.TRAININGS);
    });

    test(`If 'subjectExams' doesn't exist component shouldn't display exam type selection`, () => {
        /** Render component */
        const tree = shallow(
            <Component
                {...requiredProps}
                subjectExams={null}
            />,
        );

        /** Simulate onChange event in select-test-type text field */
        tree.find(`[data-testid='select-test-type']`)
            .simulate('change', { target: { value: ETestTypes.EXAMS } });

        /** Assert select-test-type field has value 'EXAMS' */
        expect(tree.find(`[data-testid='select-test-type']`).props().value).toBe(ETestTypes.EXAMS);

        /** Assert select-exam-type field doesn't exist */
        expect(tree.exists(`[data-testid='select-exam-type']`)).toBeFalsy();
    });

    test(`Check is select-exam label has right value`, () => {
        /** Render component */
        const tree = shallow(
            <Component
                {...requiredProps}
                subjectExams={{
                    trainings: ['foo'],
                    sessions: ['bar'],
                }}
            />,
        );

        /** Select 'EXAMS' test type */
        tree.find(`[data-testid='select-test-type']`)
            .simulate('change', { target: { value: ETestTypes.EXAMS } });

        /** Select 'TRAININGS' exam type */
        tree.find(`[data-testid='select-exam-type']`)
            .simulate('change', { target: { value: EExamTypes.TRAININGS } });

        /** Assert select-exam label has right value */
        expect(tree.find(`[data-testid='select-exam-title']`).props().children).toBe('Виберіть тренувальний варіант');

        /** Select 'SESSIONS' exam type */
        tree.find(`[data-testid='select-exam-type']`)
            .simulate('change', { target: { value: EExamTypes.SESSIONS } });

        expect(tree.find(`[data-testid='select-exam-title']`).props().children).toBe('Виберіть варіант ЗНО');
    });

    test(`Check is after exam type selection select-exam text field should has right value`, () => {
        /** Render component */
        const tree = shallow(
            <Component
                {...requiredProps}
                subjectExams={{
                    trainings: ['foo'],
                    sessions: ['bar'],
                }}
            />,
        );

        /** Select 'EXAMS' test type */
        tree.find(`[data-testid='select-test-type']`)
            .simulate('change', { target: { value: ETestTypes.EXAMS } });

        /** Select 'TRAININGS' exam type */
        tree.find(`[data-testid='select-exam-type']`)
            .simulate('change', { target: { value: EExamTypes.TRAININGS } });

        /** Assert select-exam text field has right value */
        expect(tree.find(`[data-testid='select-exam']`).props().value).toEqual('foo');

        /** Select 'SESSIONS' exam type */
        tree.find(`[data-testid='select-exam-type']`)
            .simulate('change', { target: { value: EExamTypes.SESSIONS } });

        /** Assert select-exam text field has right value */
        expect(tree.find(`[data-testid='select-exam']`).props().value).toEqual('bar');
    });

    test(`If property 'trainings' is null exam value shouldn't be displayed`, () => {
        /** Render component */
        const tree = shallow(
            <Component
                {...requiredProps}
                subjectExams={{
                    trainings: null,
                    sessions: ['bar'],
                }}
            />,
        );

        /** Select 'EXAMS' test type */
        tree.find(`[data-testid='select-test-type']`)
            .simulate('change', { target: { value: ETestTypes.EXAMS } });

        /** Select 'TRAININGS' exam type */
        tree.find(`[data-testid='select-exam-type']`)
            .simulate('change', { target: { value: EExamTypes.TRAININGS } });

        /** Assert select-exam text field is not displayed */
        expect(tree.exists(`[data-testid='select-exam']`)).toBeFalsy();
    });

    test(`If property 'sessions' is null exam value shouldn't be displayed`, () => {
        /** Render component */
        const tree = shallow(
            <Component
                {...requiredProps}
                subjectExams={{
                    trainings: ['foo'],
                    sessions: null,
                }}
            />,
        );

        /** Select 'EXAMS' test type */
        tree.find(`[data-testid='select-test-type']`)
            .simulate('change', { target: { value: ETestTypes.EXAMS } });

        /** Select 'SESSIONS' exam type */
        tree.find(`[data-testid='select-exam-type']`)
            .simulate('change', { target: { value: EExamTypes.SESSIONS } });

        /** Assert select-exam text field is not displayed */
        expect(tree.exists(`[data-testid='select-exam']`)).toBeFalsy();
    });

    test(`Check is after exam selection select-exam field has right value`, () => {
        /** Render component */
        const tree = shallow(
            <Component
                {...requiredProps}
                subjectExams={{
                    trainings: ['foo'],
                    sessions: ['bar'],
                }}
            />,
        );

        /** Select 'EXAMS' test type */
        tree.find(`[data-testid='select-test-type']`)
            .simulate('change', { target: { value: ETestTypes.EXAMS } });

        /** Select 'SESSIONS' exam type */
        tree.find(`[data-testid='select-exam-type']`)
            .simulate('change', { target: { value: EExamTypes.SESSIONS } });

        /** Simulate onChange event in select-exam text field */
        tree.find(`[data-testid='select-exam']`)
            .simulate('change', { target: { value: 'bar' } });

        /** Assert select-exam texi field has right value */
        expect(tree.find(`[data-testid='select-exam']`).props().value).toBe('bar');
    });

    test('Check is after sub-subject selection select-subject-theme has right value', () => {
        /** Render component */
        const tree = shallow(
            <Component
                {...requiredProps}
                subSubjectsData={[{
                    name: 'foo',
                    id: '123',
                }, {
                    name: 'abc',
                    id: '456',
                }]}
                subSubjectsThemes={{ foo: ['bar'], abc: ['123', '456'] }}
            />,
        );

        /** Simulate onChange event in select-test-type text field */
        tree.find(`[data-testid='select-test-type']`)
            .simulate('change', { target: { value: ETestTypes.THEMES } });

        /** Assert select-subject-theme text field has right value */
        expect(tree.find(`[data-testid='select-subject-theme']`).props().value).toBe('bar');

        /** Select 'abc' sub-subject */
        tree.find(`[data-testid='select-sub-subject']`)
            .simulate('change', { target: { value: 'abc' } });

        /** Assert select-sub-subject has value 'abc' */
        expect(tree.find(`[data-testid='select-sub-subject']`).props().value).toBe('abc');

        /** Select second element in 'abc' array */
        tree.find(`[data-testid='select-subject-theme']`)
            .simulate('change', { target: { value: '456' } });

        /** Assert select-subject-theme text field has value '456' */
        expect(tree.find(`[data-testid='select-subject-theme']`).props().value).toBe('456');

        /** Select 'foo' sub-subject */
        tree.find(`[data-testid='select-sub-subject']`)
            .simulate('change', { target: { value: 'foo' } });

        /** Assert select-subject-theme text field value is first element of 'foo' array - 'bar' */
        expect(tree.find(`[data-testid='select-subject-theme']`).props().value).toBe('bar');
    });

    test(`If 'subSubject' doesn't exist in subSubjectThemes, select-subject-theme value should be ''`, () => {
        /** Render component */
        const tree = shallow(
            <Component
                {...requiredProps}
                subSubjectsData={[{
                    name: 'foo',
                    id: '123',
                }, {
                    name: 'abc',
                    id: '456',
                }]}
                subSubjectsThemes={{ foo: ['bar'], abcd: ['123', '456'] }}
            />,
        );

        /** Simulate onChange event in select-test-type text field */
        tree.find(`[data-testid='select-test-type']`)
            .simulate('change', { target: { value: ETestTypes.THEMES } });

        /** Assert select-subject-theme text field has right value */
        expect(tree.find(`[data-testid='select-subject-theme']`).props().value).toBe('bar');

        /** Select 'abc' sub-subjec */
        tree.find(`[data-testid='select-sub-subject']`)
            .simulate('change', { target: { value: 'abc' } });

        /** Assert select-subject-theme text field has empty value */
        expect(tree.find(`[data-testid='select-subject-theme']`).props().value).toBe('');
    });

    test(`If theme or exam is not selected 'go-to-test' button should be disabled`, () => {
        /** Render component */
        const tree = shallow(
            <Component
                {...requiredProps}
                subjectThemes={['foo', 'bar']}
            />,
        );

        /** Assert 'go-to-test' button is disabled */
        expect(tree.find(`[data-testid='go-to-test']`).props().disabled).toBeTruthy();
    });

    test(`If theme or exam is selected 'go-to-test' butotn shouldn't be disabled`, () => {
        /** Render component */
        const tree = shallow(
            <Component
                {...requiredProps}
                subjectThemes={['foo', 'bar']}
                subjectExams={{
                    sessions: ['foo123'],
                    trainings: ['bar456'],
                }}
            />,
        );

        /** Select test type 'THEMES' */
        tree.find(`[data-testid='select-test-type']`)
            .simulate('change', { target: { value: ETestTypes.THEMES } });

        /** Assert 'go-to-test' button is active */
        expect(tree.find(`[data-testid='go-to-test']`).props().disabled).toBeFalsy();

        /** Select test type 'EXAMS' */
        tree.find(`[data-testid='select-test-type']`)
            .simulate('change', { target: { value: ETestTypes.EXAMS } });

        /** Select exam type */
        tree.find(`[data-testid='select-exam-type']`)
            .simulate('change', { target: { value: EExamTypes.SESSIONS } });

        /** Assert 'go-to-test' button is active */
        expect(tree.find(`[data-testid='go-to-test']`).props().disabled).toBeFalsy();

        /** Select exam type */
        tree.find(`[data-testid='select-exam-type']`)
            .simulate('change', { target: { value: EExamTypes.TRAININGS } });

        /** Assert 'go-to-test' button is active */
        expect(tree.find(`[data-testid='go-to-test']`).props().disabled).toBeFalsy();
    });

    test('Go to test button with selected theme should call push method with right args', () => {
        /** Render component */
        const tree = shallow(
            <Component
                {...requiredProps}
                subjectThemes={['foo', 'bar']}
            />,
        );

        /** Select test type 'THEMES' */
        tree.find(`[data-testid='select-test-type']`)
            .simulate('change', { target: { value: ETestTypes.THEMES } });

        /** Simulate button click */
        tree.find(`[data-testid='go-to-test']`)
            .simulate('click');

        /** Assert push method called with right args */
        expect(push).toBeCalledWith('/test-suite?subjectId=123&theme=foo&showRightDuringTest=false');
    });

    test('Go to test button with sub-subjects and themes', () => {
        /** Render component */
        const tree = shallow(
            <Component
                {...requiredProps}
                subSubjectsData={[{
                    name: 'foo',
                    id: '123',
                }, {
                    name: 'abc',
                    id: '456',
                }]}
                subSubjectsThemes={{
                    foo: ['bar123', '456'],
                    abc: ['bar', 'bbb'],
                }}
            />,
        );

        /** Select test type 'THEMES' */
        tree.find(`[data-testid='select-test-type']`)
            .simulate('change', { target: { value: ETestTypes.THEMES } });

        /** Simulate button click */
        tree.find(`[data-testid='go-to-test']`)
            .simulate('click');

        /** Assert push method called with right args */
        expect(push).toBeCalledWith('/test-suite?subjectId=123&subSubjectId=123&theme=bar123&showRightDuringTest=false');
    });

    test('Go to test button with exams', () => {
        /** Render component */
        const tree = shallow(
            <Component
                {...requiredProps}
                subjectExams={{
                    trainings: ['foo'],
                    sessions: ['bar'],
                }}
            />,
        );

        /** Select test type 'THEMES' */
        tree.find(`[data-testid='select-test-type']`)
            .simulate('change', { target: { value: ETestTypes.EXAMS } });

        /** Select exam type */
        tree.find(`[data-testid='select-exam-type']`)
            .simulate('change', { target: { value: EExamTypes.SESSIONS } });

        /** Simulate button click */
        tree.find(`[data-testid='go-to-test']`)
            .simulate('click');

        /** Assert push method called with right args */
        expect(push).toBeCalledWith('/test-suite?subjectId=123&session=bar&limitTime=false');

        /** Select exam type */
        tree.find(`[data-testid='select-exam-type']`)
            .simulate('change', { target: { value: EExamTypes.TRAININGS } });

        /** Simulate button click */
        tree.find(`[data-testid='go-to-test']`)
            .simulate('click');

        /** Assert push method called with right args */
        expect(push).toBeCalledWith('/test-suite?subjectId=123&showRightDuringTest=false&training=foo&limitTime=false');
    });
});
