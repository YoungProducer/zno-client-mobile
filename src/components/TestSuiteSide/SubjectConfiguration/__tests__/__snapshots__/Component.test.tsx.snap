// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SubjectConfigurationModal component Is match snapshot 1`] = `
<Component
  aria-describedby=""
  aria-labelledby="subject-config-title"
>
  <div
    className="makeStyles-dialogHeader-4"
  >
    <WithStyles(ForwardRef(IconButton))
      className="makeStyles-homeIcon-10"
      onClick={[Function]}
    >
      <HomeOutlinedIcon
        className="makeStyles-icon-9"
      />
    </WithStyles(ForwardRef(IconButton))>
    <WithStyles(ForwardRef(Typography))
      className="makeStyles-subjectName-6"
      variant="h6"
    >
      foo
    </WithStyles(ForwardRef(Typography))>
  </div>
  <div
    className="makeStyles-selectionBlock-5"
  >
    <WithStyles(Component)
      id="subject-config-title"
    >
      Налаштування тесту
    </WithStyles(Component)>
    <WithStyles(WithStyles(ForwardRef(DialogContent)))>
      <WithStyles(ForwardRef(FormControl))
        className="makeStyles-block-1"
        component="fieldset"
      >
        <WithStyles(ForwardRef(FormLabel))
          component="legend"
        >
          Оберіть тип тесту
        </WithStyles(ForwardRef(FormLabel))>
        <ForwardRef(RadioGroup)
          aria-label="select-test-type"
          data-testid="select-test-type"
          name="select-test-type"
          onChange={[Function]}
          value=""
        >
          <WithStyles(ForwardRef(FormControlLabel))
            control={
              <WithStyles(ForwardRef(Radio))
                color="secondary"
              />
            }
            label="Вибір теми"
            labelPlacement="end"
            value="Вибір теми"
          />
          <WithStyles(ForwardRef(FormControlLabel))
            control={
              <WithStyles(ForwardRef(Radio))
                color="secondary"
              />
            }
            label="Тренувальні варіанти ЗНО"
            labelPlacement="end"
            value="Тренувальні варіанти ЗНО"
          />
        </ForwardRef(RadioGroup)>
      </WithStyles(ForwardRef(FormControl))>
      <Component
        formControl={
          Object {
            "className": "makeStyles-checkBoxRoot-11",
          }
        }
        formControlLabel={
          Object {
            "className": "makeStyles-checkBoxLabel-12",
            "label": "Показувати правильну відповідь під час тесту",
          }
        }
        onChange={[Function]}
        value={false}
      />
    </WithStyles(WithStyles(ForwardRef(DialogContent)))>
    <WithStyles(WithStyles(ForwardRef(DialogActions)))>
      <WithStyles(ForwardRef(Button))
        className="makeStyles-button-7"
        color="primary"
        data-testid="go-to-test"
        disabled={true}
        onClick={[Function]}
        variant="contained"
      >
        Розпочати тест
      </WithStyles(ForwardRef(Button))>
      <WithStyles(ForwardRef(Button))
        className="makeStyles-button-7 makeStyles-declineButton-8"
        color="primary"
        data-testid="close-subject-conf-button"
        onClick={[Function]}
        variant="text"
      >
        Скасувати
      </WithStyles(ForwardRef(Button))>
    </WithStyles(WithStyles(ForwardRef(DialogActions)))>
  </div>
</Component>
`;
